server/src/test/java/org/solarbank/server/integration/EnergySavingControllerIT.java

```diff
@@ -0,0 +1,76 @@
+package org.solarbank.server.integration;
+
+import org.junit.jupiter.api.Test;
+import org.solarbank.server.EnergySavingController;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@AutoConfigureMockMvc
+public class EnergySavingControllerIT extends IntegrationTestBase {
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Test
+    public void energySavingController_validInput_status200() throws Exception {
```

Comment:
This is exactly the right kind of test we want to include, however there is a better and cleaner way of building your request object that you pass to `mockMvc`. Here is an example below using jackson for the object mapping (`import com.fasterxml.jackson.databind.ObjectMapper`):

```
    @Test
    public void energySavingController_validInput_status200() throws Exception {
        ObjectMapper requestMapper = new ObjectMapper();

        Location location = new Location();
        location.setLat(90.0);
        location.setLongitude(45.0);

        PanelSize panelSize = new PanelSize();
        panelSize.setHeight(2.0);
        panelSize.setWidth(3.0);

        EnergyTariff energyTariff = new EnergyTariff();
        energyTariff.setAmount(0.01);
        energyTariff.setCurrencyCode("USD");

        CalculateRequest calculateRequest = new CalculateRequest();
        calculateRequest.setLocation(location);
        calculateRequest.setPanelSize(panelSize);
        calculateRequest.setEnergyTariff(energyTariff);

        String validRequest = requestMapper.writeValueAsString(calculateRequest);

        mockMvc.perform(post("/v1.0/api/calculate")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(validRequest))
                .andExpect(status().isOk())
                .andDo(print())
                .andExpect(jsonPath("$.EnergyGenPerYear").value(1.0))
                .andExpect(jsonPath("$.EnergyGenPerMonth.January").value(0.1))
                .andExpect(jsonPath("$.EnergyGenPerMonth.February").value(0.2))
                .andExpect(jsonPath("$.SavingsPerYear.CurrencyCode").value("USD"))
                .andExpect(jsonPath("$.SavingsPerYear.Amount").value(1000.0));
    }
```

To take this to the next level you could also extract the construction of `calculateRequest` to its own method so it can be reused by your other tests. You can just call the setters on the returned object if you want to change the values to specific things in a test.

Reply:
now using jackson for object mapping and also using a separate method for constructing calculateRequest